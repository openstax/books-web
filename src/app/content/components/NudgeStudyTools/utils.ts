import { HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React from 'react';
import { useSelector } from 'react-redux';
import { useDebouncedWindowSize } from '../../../reactUtils';
import { assertDocument, remsToPx } from '../../../utils';
import { page as pageSelector } from '../../selectors';
import { hasStudyGuides, studyGuidesEnabled } from '../../studyGuides/selectors';
import {
  arrowDesktopHeight,
  arrowLeftMargin,
  arrowMobileHeight,
  arrowTopMargin,
  closeButtonDistanceFromContent,
  contentMarginTop,
  cookieNudge,
  nudgeStudyToolsMinPageLimit,
  nudgeStudyToolsShowLimit,
  nudgeStudyToolsTargetId,
  spotlightPadding,
  timeIntervalBetweenShowingNudgeInMs,
} from './constants';

interface Positions {
  arrowLeft: number;
  arrowTopOffset: number;
  closeButtonLeft: number;
  closeButtonTopOffset: number;
  contentWrapperRight: number;
  contentWrapperTopOffset: number;
  spotlightHeight: number;
  spotlightLeftOffset: number;
  spotlightTopOffset: number;
  spotlightWidth: number;
}

export const getPositions = (target: HTMLElement, isMobile: boolean, windowWidth: number): Positions => {
  const { top, left, right, height, width } = getBoundingRectOfNudgeTarget(target);
  const padding = remsToPx(spotlightPadding);
  const spotlightTopOffset = top - padding;
  const spotlightLeftOffset = left - padding;
  const spotlightHeight = height + (padding * 2);
  const spotlightWidth = width + (padding * 2);
  const arrowTopOffset = spotlightTopOffset + spotlightHeight + remsToPx(arrowTopMargin);
  const arrowLeft = spotlightLeftOffset + remsToPx(arrowLeftMargin);
  const contentWrapperTopOffset = arrowTopOffset
    + remsToPx(isMobile ? arrowMobileHeight : arrowDesktopHeight)
    + remsToPx(contentMarginTop);
  const contentWrapperRight = windowWidth - right - padding;
  const closeButtonLeft = left + spotlightWidth;
  const closeButtonTopOffset = contentWrapperTopOffset - remsToPx(closeButtonDistanceFromContent);

  return {
    arrowLeft,
    arrowTopOffset,
    closeButtonLeft,
    closeButtonTopOffset,
    contentWrapperRight,
    contentWrapperTopOffset,
    spotlightHeight,
    spotlightLeftOffset,
    spotlightTopOffset,
    spotlightWidth,
  };
};

// Target may have display set to `contents` so we are calculatings rect for children of passed target.
const getBoundingRectOfNudgeTarget = (target: HTMLElement) => {
  // Starting values depends on if we checks for Math.max or Math.min
  let top = 9999;
  let bottom = 0;
  let left = 9999;
  let right = 0;

  for (const child of Array.from(target.children)) {
    const rect = child.getBoundingClientRect();
    top = Math.min(top, rect.top);
    bottom = Math.max(bottom, rect.bottom);
    left = Math.min(left, rect.left);
    right = Math.max(right, rect.right);
  }

  return {
    height: bottom - top,
    left,
    right,
    top,
    width: right - left,
  };
};

export const usePositions = (isMobile: boolean) => {
  const [windowWidth] = useDebouncedWindowSize();
  const [positions, setPositions] = React.useState<Positions | null>(null);
  const studyGuides = useSelector(hasStudyGuides);
  const isEnabled = useSelector(studyGuidesEnabled);

  React.useEffect(() => {
    const document = assertDocument();
    const target = document.querySelector(`#${nudgeStudyToolsTargetId}`) as HTMLElement | null;
    if (isEnabled && target) {
      // Make sure that we calculate positions with body overflow set to hidden
      // because it causes scrollbar to hide which results with different positions.
      const prevOverflow = document.body.style.overflow;
      document.body.style.overflow = 'hidden';
      setPositions(getPositions(target, isMobile, windowWidth));

      // Resets to the value from before calculations. We want this style change to be handled
      // directly in the component.
      document.body.style.overflow = prevOverflow;
    }
    return () => setPositions(null);
  }, [studyGuides, isEnabled, windowWidth, isMobile]);

  return positions;
};

export const getCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.counter) || 0);
};

export const getDateCookie = () => {
  const lastShownDate = Cookies.get(cookieNudge.date);
  return lastShownDate ? new Date(lastShownDate) : undefined;
};

export const passedTimeInterval = () => {
  const lastShownDate = getDateCookie();
  return !lastShownDate
    ? true
    : (Date.now() - lastShownDate.getTime()) > timeIntervalBetweenShowingNudgeInMs
  ;
};

export const getPageCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.pageCounter) || 0);
};

export const incrementPageCounterCookie = () => {
  const counter = getPageCounterCookie();
  Cookies.set(cookieNudge.pageCounter, (counter + 1).toString());
};

export const shouldDisplayNudgeStudyTools = (): boolean => {
  const counter = getCounterCookie();
  const numberOfPagesOpenedByUser = getPageCounterCookie();

  return counter < nudgeStudyToolsShowLimit
    && passedTimeInterval()
    && (numberOfPagesOpenedByUser >= nudgeStudyToolsMinPageLimit);
};

// Set required cookies and reset opened page counter
export const setNudgeStudyToolsCookies = () => {
  const now = new Date();
  const counter = getCounterCookie();
  Cookies.set(cookieNudge.counter, (counter + 1).toString());
  Cookies.set(cookieNudge.date, now.toString());
  Cookies.remove(cookieNudge.pageCounter);
};

export const useIncrementPageCounter = () => {
  const page = useSelector(pageSelector);
  return React.useEffect(() => {
    const counter = getPageCounterCookie();
    if (page && counter <= nudgeStudyToolsMinPageLimit) { incrementPageCounterCookie(); }
  }, [page]);
};
