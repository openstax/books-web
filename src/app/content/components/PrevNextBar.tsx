import React from 'react';
import { FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import styled, { css } from 'styled-components';
import { ChevronLeft } from 'styled-icons/boxicons-regular/ChevronLeft';
import { ChevronRight } from 'styled-icons/boxicons-regular/ChevronRight';
import { textRegularLineHeight } from '../../components/Typography';
import theme from '../../theme';
import { AppState } from '../../types';
import * as select from '../selectors';
import { ArchiveTreeSection, Book } from '../types';
import ContentLink from './ContentLink';
import { contentTextStyle } from './Page';

const prevNextIconStyles = css`
  height: ${textRegularLineHeight}rem;
  width: ${textRegularLineHeight}rem;
`;

// tslint:disable-next-line:variable-name
const LeftArrow = styled(ChevronLeft)`
  ${prevNextIconStyles}
`;

// tslint:disable-next-line:variable-name
const RightArrow = styled(ChevronRight)`
  ${prevNextIconStyles}
`;

interface HidingContentLinkProps {
  book?: Book;
  page?: ArchiveTreeSection;
}
// tslint:disable-next-line:variable-name
const HidingContentLink: React.SFC<HidingContentLinkProps> = ({page, book, ...props}) =>
  page !== undefined && book !== undefined
    ? <ContentLink book={book} page={page} {...props} />
    : <span aria-hidden />;

// tslint:disable-next-line:variable-name
const BarWrapper = styled.div`
  ${contentTextStyle}
  justify-content: space-between;
  height: 4rem;
  display: flex;
  align-items: center;
  margin: 5rem auto 4rem auto;
  border-top: solid 0.1rem ${theme.color.neutral.darkest};
  border-bottom: solid 0.1rem ${theme.color.neutral.darkest};

  a {
    border: none;
  }

  ${theme.breakpoints.mobile(css`
    margin: 3.5rem auto 3.7rem auto;
    border: 0;
  `)}
`;

interface PropTypes {
  book?: Book;
  prevNext: {
    prev?: ArchiveTreeSection;
    next?: ArchiveTreeSection;
  };
}

// tslint:disable-next-line:variable-name
const PrevNextBar: React.SFC<PropTypes> = ({book, prevNext}) => <BarWrapper>
  <FormattedMessage id='i18n:prevnext:prev:aria-label'>
    {(ariaLabel: Element | string) =>
    <HidingContentLink book={book} page={prevNext.prev} aria-label={ariaLabel}>
      <LeftArrow/>
      <FormattedMessage id='i18n:prevnext:prev:text'>
        {(msg: Element | string) => msg}
      </FormattedMessage>
    </HidingContentLink>
    }
  </FormattedMessage>

  <FormattedMessage id='i18n:prevnext:next:aria-label'>
    {(ariaLabel: Element | string) =>
    <HidingContentLink book={book} page={prevNext.next} aria-label={ariaLabel}>
      <FormattedMessage id='i18n:prevnext:next:text'>
        {(msg: Element | string) => msg}
      </FormattedMessage>
      <RightArrow/>
    </HidingContentLink>
    }
  </FormattedMessage>
</BarWrapper>;

export default connect(
  (state: AppState) => ({
    book: select.book(state),
    prevNext: select.prevNextPage(state),
  })
)(PrevNextBar);
