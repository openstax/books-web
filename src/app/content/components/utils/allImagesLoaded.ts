import { Document,  Element } from '@openstax/types/lib.dom';

export default (container: Element | Document) => {
  return Promise.all(Array.from(container.querySelectorAll('img')).map((img) => img.complete
    ? Promise.resolve()
    : new Promise<void>((resolve) => {

      const previousOnload = img.onload;
      const previousOnerror = img.onerror;

      img.onload = (...args) => {
        resolve();
        if (previousOnload) {
          previousOnload.apply(img, args);
        }
      };
      img.onerror = (...args) => {
        resolve();
        if (previousOnerror) {
          previousOnerror.apply(img, args);
        }
      };
    })
  )).then((): void => undefined);
};
