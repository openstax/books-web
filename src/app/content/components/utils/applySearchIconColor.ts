import { css } from 'styled-components/macro';
import theme from '../../../theme';
import { SearchButtonColor } from '../../types';
import { toolbarIconColor } from '../constants';

export const applySearchIconColor = (props: {colorSchema: SearchButtonColor}) =>
  props.colorSchema ? css`
    color: ${theme.color.primary[props.colorSchema].foreground};

    :hover,
    :focus {
        color: ${theme.color.primary[props.colorSchema].foreground};

        ::before {
            content:"";
            display: block;
            height: 100%;
            position: absolute;
            top: 0;
            right: 0;
            min-width: 45px;
            background-color: rgba(0, 0, 0, 0.06)
            ;
        }
    }
  ` : css`
  color: ${toolbarIconColor.base};
`;
