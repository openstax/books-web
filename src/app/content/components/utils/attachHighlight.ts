import Highlighter, { Highlight, SerializedHighlight } from '@openstax/highlighter';
//import Sentry from '../../../../helpers/Sentry';

const attachHighlight = (
  highlight: Highlight | SerializedHighlight,
  highlighter: Highlighter
) => {
  highlighter.highlight(highlight);
  const result = highlighter.getHighlight(highlight.id);
  if (!result || !result.isAttached()) {
    console.log(`Highlight with id: ${highlight.id} has not been attached.`)
    // Sentry.captureException(new Error(`Highlight with id: ${highlight.id} has not been attached.`));
  }
  return result;
};

export default attachHighlight;
