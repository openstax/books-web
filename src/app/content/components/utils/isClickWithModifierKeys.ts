import { MouseEvent } from '@openstax/types/lib.dom';

export default  (e: React.MouseEvent | MouseEvent) => e.shiftKey || e.ctrlKey || e.metaKey || e.altKey;
