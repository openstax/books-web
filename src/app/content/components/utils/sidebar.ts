import { connect } from 'react-redux';
import { css, FlattenSimpleInterpolation } from 'styled-components';
import theme from '../../../theme';
import { AppState } from '../../../types';
import * as selectors from '../../selectors';
import { State } from '../../types';

export const isOpenConnector = connect((state: AppState) => ({isOpen: selectors.tocOpen(state)}));

export const styleWhenSidebarClosed = (closedStyle: FlattenSimpleInterpolation) => css`
  ${(props: {isOpen: State['tocOpen']}) => props.isOpen === null && theme.breakpoints.mobile(closedStyle)}
  ${(props: {isOpen: State['tocOpen']}) => props.isOpen === false && closedStyle}
`;
