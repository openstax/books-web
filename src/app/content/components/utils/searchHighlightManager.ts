import Highlighter from '@openstax/highlighter';
import { SearchResultHit } from '@openstax/open-search-client';
import { HTMLElement } from '@openstax/types/lib.dom';
import flow from 'lodash/fp/flow';
import isEqual from 'lodash/fp/isEqual';
import { scrollTo } from '../../../utils';
import { SelectedResult } from '../../search/types';
import { highlightResults } from '../../search/utils';
import allImagesLoaded from './allImagesLoaded';

interface Services {
  highlighter: Highlighter;
  container: HTMLElement;
  searchResults?: SearchResultHit[];
  searchResultMap: ReturnType<typeof highlightResults>;
  selected?: SelectedResult | null;
}

const scrollToSearch = ({container, searchResultMap}: Services, selected: SelectedResult) => {
  const elementHighlights = searchResultMap.find((map) => isEqual(map.result, selected.result));
  const selectedHighlights = elementHighlights && elementHighlights.highlights[selected.highlight];
  const firstSelectedHighlight = selectedHighlights && selectedHighlights[0];

  if (firstSelectedHighlight) {
    firstSelectedHighlight.focus();
    allImagesLoaded(container).then(
      () => scrollTo(firstSelectedHighlight.elements[0])
    );
  }
};

const updateResults = (services: Services, searchResults: SearchResultHit[]): Services => {
  if (services.searchResults === searchResults) {
    return services;
  }

  services.highlighter.eraseAll();

  const searchResultMap = highlightResults(services.highlighter, searchResults);

  return {...services, searchResultMap};
};

const selectResult = (services: Services, selected: SelectedResult | null): Services => {
  services.highlighter.clearFocus();

  if (selected && services.selected !== selected) {
    scrollToSearch(services, selected);
  }

  return {...services, selected};
};

const searchHighlightManager = (initialServices: Services) =>
  (searchResults: SearchResultHit[], selected: SelectedResult | null) => {
    const newServices = flow(
      (services) => updateResults(services, searchResults),
      (services) => selectResult(services, selected)
    )(initialServices);

    return searchHighlightManager(newServices);
  };

export default (container: HTMLElement) => {
  const services = {
    container,
    highlighter: new Highlighter(container, {
      className: 'search-highlight',
    }),
    searchResultMap: [],
  };

  return searchHighlightManager(services);
};

export const stubManager: ReturnType<typeof searchHighlightManager> = () => stubManager;
