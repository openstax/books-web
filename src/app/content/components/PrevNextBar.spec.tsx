import React from 'react';
import renderer from 'react-test-renderer';
import { book as archiveBook, shortPage } from '../../../test/mocks/archiveLoader';
import { mockCmsBook } from '../../../test/mocks/osWebLoader';
import { formatBookData } from '../utils';
import { PrevNextBar } from './PrevNextBar';

const book = formatBookData(archiveBook, mockCmsBook);

describe('PrevNextBar', () => {

  it('renders `null` with no page or book', () => {
    const component = renderer.create(<PrevNextBar />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when you pass a page and book', () => {
    const component = renderer.create(<PrevNextBar page={shortPage} book={book} />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders `null` when passed a page that isn\'t in the book tree', () => {
    const pageNotInTree = {...shortPage, id: 'asdfasdfasd'};
    const component = renderer.create(<PrevNextBar page={pageNotInTree} book={book} />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});
