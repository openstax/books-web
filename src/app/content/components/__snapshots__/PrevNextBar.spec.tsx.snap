// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrevNextBar renders \`null\` when passed a page that isn't in the book tree 1`] = `
.c2 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c1 {
  height: 2.5rem;
  width: 2.5rem;
}

.c0 {
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 1.6rem;
  line-height: 2.5rem;
  max-width: 57rem;
  margin: 0 auto;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 5rem auto 4rem auto;
  border-top: solid 0.1rem #e5e5e5;
  border-bottom: solid 0.1rem #e5e5e5;
}

.c0 a {
  color: #027EB5;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-bottom: #027EB5 solid 0.02em;
}

.c0 a:hover {
  color: #0064A0;
}

.c0 a {
  border: none;
}

@media (max-width:64em) {
  .c0 {
    margin: 3.5rem auto 3.7rem auto;
    border: 0;
  }
}

<div
  className="c0"
>
  <span
    aria-hidden={true}
  />
  <a
    aria-label="Next Page"
    href="books/book-slug-1/pages/test-page-1"
    onClick={[Function]}
  >
    Next
    <svg
      aria-hidden="true"
      className="c1 c2"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M10.707 17.707L16.414 12l-5.707-5.707-1.414 1.414L13.586 12l-4.293 4.293z"
      />
    </svg>
  </a>
</div>
`;

exports[`PrevNextBar renders \`null\` with no page or book 1`] = `null`;

exports[`PrevNextBar renders correctly when you pass a page and book 1`] = `
.c2 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c1 {
  height: 2.5rem;
  width: 2.5rem;
}

.c3 {
  height: 2.5rem;
  width: 2.5rem;
}

.c0 {
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 1.6rem;
  line-height: 2.5rem;
  max-width: 57rem;
  margin: 0 auto;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 5rem auto 4rem auto;
  border-top: solid 0.1rem #e5e5e5;
  border-bottom: solid 0.1rem #e5e5e5;
}

.c0 a {
  color: #027EB5;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-bottom: #027EB5 solid 0.02em;
}

.c0 a:hover {
  color: #0064A0;
}

.c0 a {
  border: none;
}

@media (max-width:64em) {
  .c0 {
    margin: 3.5rem auto 3.7rem auto;
    border: 0;
  }
}

<div
  className="c0"
>
  <a
    href="books/book-slug-1/pages/1-test-page-3"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="c1 c2"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M13.293 6.293L7.586 12l5.707 5.707 1.414-1.414L10.414 12l4.293-4.293z"
      />
    </svg>
    Previous
  </a>
  <a
    aria-label="Next Page"
    href="books/book-slug-1/pages/1-test-page-5"
    onClick={[Function]}
  >
    Next
    <svg
      aria-hidden="true"
      className="c3 c2"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M10.707 17.707L16.414 12l-5.707-5.707-1.414 1.414L13.586 12l-4.293 4.293z"
      />
    </svg>
  </a>
</div>
`;

exports[`PrevNextBar renders only \`next\` element correctly 1`] = `
.c2 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c1 {
  height: 2.5rem;
  width: 2.5rem;
}

.c0 {
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 1.6rem;
  line-height: 2.5rem;
  max-width: 57rem;
  margin: 0 auto;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 5rem auto 4rem auto;
  border-top: solid 0.1rem #e5e5e5;
  border-bottom: solid 0.1rem #e5e5e5;
}

.c0 a {
  color: #027EB5;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-bottom: #027EB5 solid 0.02em;
}

.c0 a:hover {
  color: #0064A0;
}

.c0 a {
  border: none;
}

@media (max-width:64em) {
  .c0 {
    margin: 3.5rem auto 3.7rem auto;
    border: 0;
  }
}

<div
  className="c0"
>
  <span
    aria-hidden={true}
  />
  <a
    aria-label="Next Page"
    href="books/book-slug-1/pages/1-test-page-2"
    onClick={[Function]}
  >
    Next
    <svg
      aria-hidden="true"
      className="c1 c2"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M10.707 17.707L16.414 12l-5.707-5.707-1.414 1.414L13.586 12l-4.293 4.293z"
      />
    </svg>
  </a>
</div>
`;

exports[`PrevNextBar renders only \`prev\` element correctly 1`] = `
.c2 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c1 {
  height: 2.5rem;
  width: 2.5rem;
}

.c0 {
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 1.6rem;
  line-height: 2.5rem;
  max-width: 57rem;
  margin: 0 auto;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 5rem auto 4rem auto;
  border-top: solid 0.1rem #e5e5e5;
  border-bottom: solid 0.1rem #e5e5e5;
}

.c0 a {
  color: #027EB5;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-bottom: #027EB5 solid 0.02em;
}

.c0 a:hover {
  color: #0064A0;
}

.c0 a {
  border: none;
}

@media (max-width:64em) {
  .c0 {
    margin: 3.5rem auto 3.7rem auto;
    border: 0;
  }
}

<div
  className="c0"
>
  <a
    href="books/book-slug-1/pages/1-test-page-6"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="c1 c2"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M13.293 6.293L7.586 12l5.707 5.707 1.414-1.414L10.414 12l4.293-4.293z"
      />
    </svg>
    Previous
  </a>
  <span
    aria-hidden={true}
  />
</div>
`;
